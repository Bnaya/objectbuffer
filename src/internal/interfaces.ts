import type { Heap } from "../structsGenerator/consts";
import type { TransactionalAllocator } from "./TransactionalAllocator";

export interface MemoryStats {
  available: number;
  used: number;
  total: number;
  top: number;
}

export interface GlobalCarrier {
  allocator: TransactionalAllocator;
  heap: Heap;
}

export interface ExternalArgs {
  /**
   * The load factor of the internal hash maps in use
   * @default 0.75
   */
  readonly hashMapLoadFactor: number;
  /**
   * Minimal initial buckets count of internal hash maps in use
   * @default 8
   */
  readonly hashMapMinInitialCapacity: number;
  /**
   * Allocate additional memory for array pointers,
   * will prevent the reallocation and copy when array is getting bigger
   * @default 0
   */
  readonly arrayAdditionalAllocation: number;
}

export interface ObjectBufferSettings {
  /**
   * The load factor of the internal hash maps in use
   * @default 0.75
   */
  readonly hashMapLoadFactor?: number;
  /**
   * Minimal initial buckets count of internal hash maps in use
   * @default 8
   */
  readonly hashMapMinInitialCapacity?: number;
  /**
   * Allocate additional memory for array pointers,
   * will prevent the reallocation and copy when array is getting bigger
   * @default 0
   */
  readonly arrayAdditionalAllocation?: number;
}

export interface InternalAPI {
  getExternalArgs(): ExternalArgs;
  getCarrier(): Readonly<GlobalCarrier>;
  replaceCarrierContent(carrier: GlobalCarrier): void;
  getEntryPointer(): number;
  destroy(): number;
}
