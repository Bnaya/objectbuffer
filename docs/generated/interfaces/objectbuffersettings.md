[@bnaya/objectbuffer](../README.md) / ObjectBufferSettings

# Interface: ObjectBufferSettings

## Table of contents

### Properties

- [arrayAdditionalAllocation](objectbuffersettings.md#arrayadditionalallocation)
- [hashMapLoadFactor](objectbuffersettings.md#hashmaploadfactor)
- [hashMapMinInitialCapacity](objectbuffersettings.md#hashmapmininitialcapacity)

## Properties

### arrayAdditionalAllocation

• `Optional` `Readonly` **arrayAdditionalAllocation**: `number`

Allocate additional memory for array pointers,
will prevent the reallocation and copy when array is getting bigger

**`default`** 0

#### Defined in

[internal/interfaces.ts:51](https://github.com/Bnaya/objectbuffer/blob/266152d/src/internal/interfaces.ts#L51)

___

### hashMapLoadFactor

• `Optional` `Readonly` **hashMapLoadFactor**: `number`

The load factor of the internal hash maps in use

**`default`** 0.75

#### Defined in

[internal/interfaces.ts:40](https://github.com/Bnaya/objectbuffer/blob/266152d/src/internal/interfaces.ts#L40)

___

### hashMapMinInitialCapacity

• `Optional` `Readonly` **hashMapMinInitialCapacity**: `number`

Minimal initial buckets count of internal hash maps in use

**`default`** 8

#### Defined in

[internal/interfaces.ts:45](https://github.com/Bnaya/objectbuffer/blob/266152d/src/internal/interfaces.ts#L45)
