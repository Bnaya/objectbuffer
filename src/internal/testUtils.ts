import { FIRST_FREE_BYTE_POINTER_TO_POINTER } from "./consts";

export function arrayBuffer2HexArray(
  buffer: ArrayBuffer,
  withByteNumber = false
) {
  if (withByteNumber) {
    return Array.prototype.map.call(
      new Uint8Array(buffer),
      (x: number, index) => `${index}:${x.toString(16).padStart(4, "0x")}`
    );
  }
  return Array.prototype.map.call(new Uint8Array(buffer), (x: number) =>
    x.toString(16).padStart(4, "0x")
  );
}

export function getFirstFreeByte(arrayBuffer: ArrayBuffer) {
  return new DataView(arrayBuffer).getUint32(
    FIRST_FREE_BYTE_POINTER_TO_POINTER
  );
}
