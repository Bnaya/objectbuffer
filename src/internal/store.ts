import { ENTRY_TYPE, isPrimitiveEntryType } from "./entry-types";
import { Entry, ObjectPropEntry, primitive } from "./interfaces";
import { arrayBufferCopyTo, isPrimitive, primitiveValueToEntry } from "./utils";
import { ExternalArgs } from "./interfaces";

export function initializeArrayBuffer(arrayBuffer: ArrayBuffer) {
  const dataView = new DataView(arrayBuffer);

  // global lock
  dataView.setInt32(0, 0);

  // End of data pointer / first free byte
  dataView.setUint32(8, 24);

  // first entry pointer
  dataView.setUint32(16, 24);

  return dataView;
}

export function writeEntry(
  externalArgs: ExternalArgs,
  dataView: DataView,
  startingCursor: number,
  entry: Entry
) {
  let cursor = startingCursor;

  // let writtenDataSizeInBytes = 0;

  // write type
  // undo on throw ?
  dataView.setUint8(cursor, entry.type);
  cursor += Uint8Array.BYTES_PER_ELEMENT;

  switch (entry.type) {
    case ENTRY_TYPE.UNDEFINED:
      break;

    case ENTRY_TYPE.NULL:
      break;

    case ENTRY_TYPE.BOOLEAN:
      dataView.setUint8(cursor, entry.value ? 1 : 0);
      cursor += Uint8Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.NUMBER:
      dataView.setFloat64(cursor, entry.value);
      cursor += Float64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.STRING:
      // eslint-disable-next-line no-case-declarations
      const encodedString: Uint8Array = externalArgs.textEncoder.encode(
        entry.value
      );
      dataView.setUint16(cursor, encodedString.byteLength);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      dataView.setUint16(cursor, entry.allocatedBytes);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      for (let i = 0; i < encodedString.length; i++) {
        dataView.setUint8(cursor, encodedString[i]);
        cursor += Uint8Array.BYTES_PER_ELEMENT;
      }

      cursor += entry.allocatedBytes;

      break;

    case ENTRY_TYPE.BIGINT:
      dataView.setBigInt64(cursor, entry.value);
      cursor += BigInt64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.UBIGINT:
      dataView.setBigUint64(cursor, entry.value);
      cursor += BigUint64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.OBJECT:
      dataView.setUint32(cursor, entry.value);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.OBJECT_PROP:
      // eslint-disable-next-line no-case-declarations
      const encodedStringKey: Uint8Array = externalArgs.textEncoder.encode(
        entry.value.key
      );
      dataView.setUint16(cursor, encodedStringKey.byteLength);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      for (let i = 0; i < encodedStringKey.length; i++) {
        dataView.setUint8(cursor, encodedStringKey[i]);
        cursor += Uint8Array.BYTES_PER_ELEMENT;
      }

      dataView.setUint32(cursor, entry.value.value);
      cursor += Uint32Array.BYTES_PER_ELEMENT;

      dataView.setUint32(cursor, entry.value.next);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.ARRAY:
      dataView.setUint32(cursor, entry.value);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      dataView.setUint32(cursor, entry.length);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      dataView.setUint32(cursor, entry.allocatedLength);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    default:
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      throw new Error(ENTRY_TYPE[entry.type] + " Not implemented yet");
  }

  return cursor - startingCursor;
}

export function appendEntry(
  externalArgs: ExternalArgs,
  dataView: DataView,
  entry: Entry
) {
  // End of data pointer
  const firstFreeByte = dataView.getUint32(8);

  const written = writeEntry(externalArgs, dataView, firstFreeByte, entry);
  dataView.setUint32(8, firstFreeByte + written);

  return {
    start: firstFreeByte,
    length: written
  };
}

export function readEntry(
  externalArgs: ExternalArgs,
  dataView: DataView,
  startingCursor: number
): [Entry, number] {
  let cursor = startingCursor;

  const entryType: ENTRY_TYPE = dataView.getUint8(cursor);
  cursor += Uint8Array.BYTES_PER_ELEMENT;

  const entry: any = {
    type: entryType,
    value: undefined as any
  };

  // let writtenDataSizeInBytes = 0;

  switch (entryType) {
    case ENTRY_TYPE.UNDEFINED:
      break;

    case ENTRY_TYPE.NULL:
      break;

    case ENTRY_TYPE.BOOLEAN:
      entry.value = dataView.getUint8(cursor) !== 0;
      cursor += Uint8Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.NUMBER:
      entry.value = dataView.getFloat64(cursor);
      cursor += Float64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.STRING:
      // eslint-disable-next-line no-case-declarations
      const stringLength = dataView.getUint16(cursor);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      entry.allocatedBytes = dataView.getUint16(cursor);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      // this wrapping is needed until:
      // https://github.com/whatwg/encoding/issues/172
      // eslint-disable-next-line no-case-declarations
      const tempAB = new ArrayBuffer(stringLength);
      arrayBufferCopyTo(dataView.buffer, cursor, stringLength, tempAB, 0);

      entry.value = externalArgs.textDecoder.decode(tempAB);

      cursor += stringLength;

      break;

    case ENTRY_TYPE.BIGINT:
      entry.value = dataView.getBigInt64(cursor);
      cursor += BigInt64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.UBIGINT:
      entry.value = dataView.getBigUint64(cursor);
      cursor += BigUint64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.OBJECT:
      entry.value = dataView.getUint32(cursor);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.OBJECT_PROP:
      // eslint-disable-next-line no-case-declarations
      const keyStringLength = dataView.getUint16(cursor);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      // this wrapping is needed until:
      // https://github.com/whatwg/encoding/issues/172
      // eslint-disable-next-line no-case-declarations
      const tempAB2 = new ArrayBuffer(keyStringLength);
      arrayBufferCopyTo(dataView.buffer, cursor, keyStringLength, tempAB2, 0);

      // eslint-disable-next-line no-case-declarations
      const objectPropsValue: ObjectPropEntry["value"] = {
        key: externalArgs.textDecoder.decode(tempAB2),
        value: dataView.getUint32(cursor + keyStringLength),
        next: dataView.getUint32(
          cursor + keyStringLength + Uint32Array.BYTES_PER_ELEMENT
        )
      };

      cursor +=
        keyStringLength +
        Uint32Array.BYTES_PER_ELEMENT +
        Uint32Array.BYTES_PER_ELEMENT;

      entry.value = objectPropsValue;
      break;

    case ENTRY_TYPE.ARRAY:
      entry.value = dataView.getUint32(cursor);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      entry.length = dataView.getUint32(cursor);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      entry.allocatedLength = dataView.getUint32(cursor);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    default:
      throw new Error(ENTRY_TYPE[entryType] + " Not implemented yet");
  }

  return [entry, cursor - startingCursor];
}

export function reserveMemory(dataView: DataView, length: number) {
  const firstFreeByte = dataView.getUint32(8);

  dataView.setUint32(8, firstFreeByte + length);

  return firstFreeByte;
}

export function canSaveValueIntoEntry(
  externalArgs: ExternalArgs,
  entryA: Entry,
  value: primitive
) {
  const typeofTheValue = typeof value;

  if (entryA.type === ENTRY_TYPE.BOOLEAN && typeofTheValue === "boolean") {
    return true;
  }

  // number & bigint 64 are the same size
  if (
    (entryA.type === ENTRY_TYPE.BIGINT ||
      entryA.type === ENTRY_TYPE.UBIGINT ||
      entryA.type === ENTRY_TYPE.NUMBER) &&
    (typeofTheValue === "bigint" || typeofTheValue === "number")
  ) {
    return true;
  }

  if (
    entryA.type === ENTRY_TYPE.STRING &&
    typeofTheValue === "string" &&
    entryA.allocatedBytes >= externalArgs.textEncoder.encode(value).length
  ) {
    return true;
  }

  if (
    ((entryA.type === ENTRY_TYPE.NULL ||
      entryA.type === ENTRY_TYPE.UNDEFINED) &&
      value === undefined) ||
    value === null
  ) {
    return true;
  }

  return false;
}

export function overwriteEntryIfPossible(
  externalArgs: ExternalArgs,
  dataView: DataView,
  pointerToExistingEntry: number,
  value: any
) {
  if (isPrimitive(value)) {
    const existingValueEntry = readEntry(
      externalArgs,
      dataView,
      pointerToExistingEntry
    );

    if (canSaveValueIntoEntry(externalArgs, existingValueEntry[0], value)) {
      const stringAllocatedBytes =
        existingValueEntry[0].type === ENTRY_TYPE.STRING
          ? existingValueEntry[0].allocatedBytes
          : 0;

      const newEntry = primitiveValueToEntry(
        externalArgs,
        value,
        stringAllocatedBytes
      );

      writeEntry(externalArgs, dataView, pointerToExistingEntry, newEntry);
      return true;
    }

    return false;
  }

  return false;
}
