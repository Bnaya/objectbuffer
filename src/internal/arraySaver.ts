import { appendEntry } from "./store";
import { ENTRY_TYPE } from "./entry-types";
import { ArrayEntry, ExternalArgs, GlobalCarrier } from "./interfaces";
import { saveValue } from "./saveValue";

export function arraySaver(
  externalArgs: ExternalArgs,
  carrier: GlobalCarrier,
  referencedPointers: number[],
  arrayToSave: Array<any>
) {
  let memoryForPointersCursor = carrier.allocator.calloc(
    (arrayToSave.length + externalArgs.arrayAdditionalAllocation) *
      Uint32Array.BYTES_PER_ELEMENT
  );

  const arrayStartEntry: ArrayEntry = {
    type: ENTRY_TYPE.ARRAY,
    refsCount: 1,
    value: memoryForPointersCursor,
    allocatedLength:
      arrayToSave.length + externalArgs.arrayAdditionalAllocation,
    length: arrayToSave.length
  };

  for (const item of arrayToSave) {
    const rOfValue = saveValue(externalArgs, carrier, referencedPointers, item);

    carrier.uint32[
      memoryForPointersCursor / Uint32Array.BYTES_PER_ELEMENT
    ] = rOfValue;

    memoryForPointersCursor += Uint32Array.BYTES_PER_ELEMENT;
  }

  const arrayEntryAppendResult = appendEntry(
    externalArgs,
    carrier,
    arrayStartEntry
  );

  return arrayEntryAppendResult;
}
