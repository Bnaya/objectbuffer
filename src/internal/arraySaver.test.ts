/* eslint-env jest */

import { initializeArrayBuffer } from "./store";
import * as util from "util";
import { arrayBuffer2HexArray } from "./testUtils";
import { arraySaver } from "./arraySaver";
import { ExternalArgs } from "./interfaces";

describe("arraySaver tests", () => {
  const externalArgs: ExternalArgs = {
    textEncoder: new util.TextEncoder(),
    textDecoder: new util.TextDecoder(),
    arrayAdditionalAllocation: 0,
    minimumStringAllocation: 0
  };

  describe("arraySaver - general", () => {
    test("arraySaver", () => {
      const arrayBuffer = new ArrayBuffer(80);

      const dataView = new DataView(arrayBuffer);
      initializeArrayBuffer(arrayBuffer);

      const arrayToSave = [1, 2, 3];

      const saverOutput = arraySaver(externalArgs, dataView, arrayToSave);

      expect(saverOutput).toMatchInlineSnapshot(`
        Object {
          "length": 52,
          "start": 63,
        }
      `);

      expect(arrayBuffer2HexArray(arrayBuffer, true)).toMatchSnapshot();
    });
  });
});
