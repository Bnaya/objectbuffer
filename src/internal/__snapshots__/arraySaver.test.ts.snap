// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arraySaver tests arraySaver - general arraySaver: after array save 1`] = `
Array [
  "0:0x00",
  "1:0x00",
  "2:0x00",
  "3:0x00",
  "4:0x00",
  "5:0x00",
  "6:0x00",
  "7:0x08",
  "8:0x00",
  "9:0x00",
  "10:0x00",
  "11:0x00",
  "12:0x00",
  "13:0x00",
  "14:0x00",
  "15:0x00",
  "16:0xe8",
  "17:0x00",
  "18:0x00",
  "19:0x00",
  "20:0x00",
  "21:0x01",
  "22:0x00",
  "23:0x00",
  "24:0x00",
  "25:0x00",
  "26:0x00",
  "27:0x00",
  "28:0xd0",
  "29:0x00",
  "30:0x00",
  "31:0x00",
  "32:0x68",
  "33:0x00",
  "34:0x00",
  "35:0x00",
  "36:0x00",
  "37:0x00",
  "38:0x00",
  "39:0x00",
  "40:0x00",
  "41:0x00",
  "42:0x00",
  "43:0x90",
  "44:0x00",
  "45:0x00",
  "46:0x00",
  "47:0xa8",
  "48:0x00",
  "49:0x00",
  "50:0x00",
  "51:0xc0",
  "52:0x00",
  "53:0x00",
  "54:0x00",
  "55:0x00",
  "56:0x00",
  "57:0x00",
  "58:0x00",
  "59:0x00",
  "60:0x00",
  "61:0x00",
  "62:0x00",
  "63:0x00",
  "64:0x00",
  "65:0x00",
  "66:0x00",
  "67:0x00",
  "68:0x00",
  "69:0x00",
  "70:0x00",
  "71:0x00",
  "72:0x00",
  "73:0x00",
  "74:0x00",
  "75:0x00",
  "76:0x00",
  "77:0x00",
  "78:0x00",
  "79:0x00",
  "80:0x00",
  "81:0x00",
  "82:0x00",
  "83:0x00",
  "84:0x00",
  "85:0x00",
  "86:0x00",
  "87:0x00",
  "88:0x00",
  "89:0x00",
  "90:0x00",
  "91:0x00",
  "92:0x00",
  "93:0x00",
  "94:0x00",
  "95:0x00",
  "96:0x00",
  "97:0x00",
  "98:0x00",
  "99:0x00",
  "100:0x00",
  "101:0x00",
  "102:0x00",
  "103:0x00",
  "104:0x00",
  "105:0x00",
  "106:0x00",
  "107:0x00",
  "108:0x00",
  "109:0x00",
  "110:0x00",
  "111:0x00",
  "112:0x00",
  "113:0x00",
  "114:0x00",
  "115:0x00",
  "116:0x00",
  "117:0x00",
  "118:0x00",
  "119:0x00",
  "120:0x00",
  "121:0x00",
  "122:0x00",
  "123:0x00",
  "124:0x00",
  "125:0x00",
  "126:0x00",
  "127:0x00",
  "128:0x00",
  "129:0x00",
  "130:0x00",
  "131:0x00",
  "132:0x00",
  "133:0x00",
  "134:0x00",
  "135:0x00",
  "136:0x18",
  "137:0x00",
  "138:0x00",
  "139:0x00",
  "140:0x20",
  "141:0x00",
  "142:0x00",
  "143:0x00",
  "144:0x02",
  "145:0x3f",
  "146:0xf0",
  "147:0x00",
  "148:0x00",
  "149:0x00",
  "150:0x00",
  "151:0x00",
  "152:0x00",
  "153:0x00",
  "154:0x00",
  "155:0x00",
  "156:0x00",
  "157:0x00",
  "158:0x00",
  "159:0x00",
  "160:0x18",
  "161:0x00",
  "162:0x00",
  "163:0x00",
  "164:0x88",
  "165:0x00",
  "166:0x00",
  "167:0x00",
  "168:0x02",
  "169:0x40",
  "170:0x00",
  "171:0x00",
  "172:0x00",
  "173:0x00",
  "174:0x00",
  "175:0x00",
  "176:0x00",
  "177:0x00",
  "178:0x00",
  "179:0x00",
  "180:0x00",
  "181:0x00",
  "182:0x00",
  "183:0x00",
  "184:0x18",
  "185:0x00",
  "186:0x00",
  "187:0x00",
  "188:0xa0",
  "189:0x00",
  "190:0x00",
  "191:0x00",
  "192:0x02",
  "193:0x40",
  "194:0x08",
  "195:0x00",
  "196:0x00",
  "197:0x00",
  "198:0x00",
  "199:0x00",
  "200:0x00",
  "201:0x00",
  "202:0x00",
  "203:0x00",
  "204:0x00",
  "205:0x00",
  "206:0x00",
  "207:0x00",
  "208:0x18",
  "209:0x00",
  "210:0x00",
  "211:0x00",
  "212:0xb8",
  "213:0x00",
  "214:0x00",
  "215:0x00",
  "216:0x09",
  "217:0x01",
  "218:0x00",
  "219:0x00",
  "220:0x00",
  "221:0x28",
  "222:0x00",
  "223:0x00",
  "224:0x00",
  "225:0x03",
  "226:0x00",
  "227:0x00",
  "228:0x00",
  "229:0x17",
  "230:0x00",
  "231:0x00",
  "232:0x00",
  "233:0x00",
  "234:0x00",
  "235:0x00",
  "236:0x00",
  "237:0x00",
  "238:0x00",
  "239:0x00",
  "240:0x00",
  "241:0x00",
  "242:0x00",
  "243:0x00",
  "244:0x00",
  "245:0x00",
  "246:0x00",
  "247:0x00",
  "248:0x00",
  "249:0x00",
  "250:0x00",
  "251:0x00",
  "252:0x00",
  "253:0x00",
  "254:0x00",
  "255:0x00",
]
`;
